### DISCUSSION TITLE
Constrained conformation generation

### DISCUSSION THREAD

**Discussion by danielzaidman** [rdkit/rdkit#4663] [Q&A]
Dear RDKit community.
My question is regarding constrained conformation generation, with two fixed substructures.
The way I've been doing it until now, is by using AllChem.EmbedMolecule, with coordMap being (some of) the fixed part's atoms. This works in most cases, but in a few rare cases it gives funny incorrect conformation for the variable part. One example I came across is a molecule with the following substructure: c1ccccc(C=CC(=O)N)1.
In some cases the conformation is valid, while in others, the bond angles between the carbonyl carbon and the carbon next to it is completely wrong, sometimes even 180 deg. I was wondering if anyone has come across this problem.

Also, thinking of a solution, I thought of replacing the EmbedMolecule with ConstrainedEmbed. However, this is not straightforward, since I have two "cores" which are disconnected, so the cmap approach is easier in a way.
Also, since most of the molecules are completely fine, as well as most conformations, and even most of the molecule within the faulty conformations, I'm guessing that this is an inner problem with the way RDKit's chemistry rules, something that ConstrainedEmbed won't fix.

Anyway, I'll be happy to hear any advice, and/or plain sympathy :)
I wish you all well-generated conformers!
Best, Daniel.

**Comment by danielzaidman** [comment]
Just to add to mu suspicions, I looked into the ConstrainedEmbed implementation, and it seems like it's doing exactly what I did, using the normal Embed function, with cmap. So unless there's something I'm totally missing, I would say this is a chemistry "mishap" by RDKit, which I don't have the time (and probably not the skills) to explore. I have seen this quite a lot of times, although it's still rare. If any of you would like to look more into this, I'd be glad to chat about it.

**Reply by fdaqin** [reply]
Hello, thanks for the disscusion. I have a problem with ConstrainedEmbed, and I can now use this to generate conformations based on one substructure but not two substructures, do you know how to do it? Thanks!

**Reply by danielzaidman** [reply]
Hi. So as far as I'm aware, there is no such option in the normal use of this function. I did something similar (or even the same) with a work around, in the PRosettaC paper, and the code is available at: https://github.com/LondonLab/PRosettaC.
Feel free to take a look. It's in the function "GenConstConf" within https://github.com/LondonLab/PRosettaC/blob/master/protac_lib.py

**Reply by greglandrum** [reply]
> I would say this is a chemistry "mishap" by RDKit, which I don't have the time (and probably not the skills) to explore. I have seen this quite a lot of times, although it's still rare. If any of you would like to look more into this, I'd be glad to chat about it.

Sorry I missed this older discussion. 
We added a flag (which is enabled by default) to force transoid geometries for amides, esters, and related structures in the 2021.03 release of the RDKit (https://github.com/rdkit/rdkit/pull/3794). If you are still observing bad structures with a release which >= 2021.03.1, please share the structure here (or go ahead and create a bug report for it).

**Reply by fdaqin** [reply]
> Hi. So as far as I'm aware, there is no such option in the normal use of this function. I did something similar (or even the same) with a work around, in the PRosettaC paper, and the code is available at: https://github.com/LondonLab/PRosettaC. Feel free to take a look. It's in the function "GenConstConf" within https://github.com/LondonLab/PRosettaC/blob/master/protac_lib.py

Thanks a lot! I'll look into these codes. Currently, I get a triangle-bounds smoothing error when generating conformations using two substructures as constraints. I tried adding ignoreSmoothingFailures=True flag (default is False), the EmbedMolecule or ConstrainedEmbed command can run normally and generate conformations, but using this parameter may lead to some unreasonable structures indeed. I would like to know how you would think about this flag. Thanks so much！

**Reply by lonngxiang** [reply]
> Hi. So as far as I'm aware, there is no such option in the normal use of this function. I did something similar (or even the same) with a work around, in the PRosettaC paper, and the code is available at: https://github.com/LondonLab/PRosettaC. Feel free to take a look. It's in the function "GenConstConf" within https://github.com/LondonLab/PRosettaC/blob/master/protac_lib.py
@fdaqin 
So does it work out? Well, the function "GenConstConf" is basically -1, and it doesn't produce any valid results

**Reply by lonngxiang** [reply]
> transoid
@greglandrum 
In function AllChem.EmbedMolecule；
Are there any parameters that can be disabled， let not force transoid geometries for amides, esters, and related structures

**Reply by fdaqin** [reply]
@lonngxiang I haven't run the above codes, but adding 'ignoreSmoothingFailures=True' can improve the possibility of conformation generation. you can have a try.
